### ISSUE TITLE
Add Db Migration, preferably Automatic

### ISSUE THREAD

**Issue by ericbusboom** [CivicKnowledge/ambry#91]
The prior version of Amby had automatic database migration for Sqlite, implemented in ambry/database/sqlite.py:686. The code used the user pragma to store the schema version, and when the version changed, it would run SQL commands. It worked really well, but only for Sqlite. 

Ambry 1.0 should have a similar system to detect when the schema version doesn't match the code version and automatically perform and upgrade, at at least notify the user that and upgrade needs to be performed. 

In the past, all of the migrations have involved additions, primarily additional columns. Simple migrations should be managed automatically, although more complex migrations may be deferred to a manual process. If the process is manual, there should be an ambry cli command to execute it.

**Comment by nmb10** [comment]
Done. See description here - https://github.com/CivicKnowledge/ambry/blob/ambry1.0/doc/DEVEL-README.md#to-change-models-fields-and-appropriate-database-table-aka-migration

**Comment by ericbusboom** [comment]
> On Aug 1, 2015, at 9:43 AM, Kazbek <notifications@github.com> wrote:
> 
> Done. See description here - https://github.com/CivicKnowledge/ambry/blob/ambry1.0/doc/DEVEL-README.md#to-change-models-fields-and-appropriate-database-table-aka-migration

The use case of needing credentials for a postgres database is probably pretty general, so the configuration should support production databases too. Here’s a way to do that.

The account entry for this example could be something like: 

	postgres_127.0.0.1_ambry:
		user: ambry
		password: secret

( The account file is intended for holding only credentials you want to keep secret, even though it currently isn’t really good at keeping things secret. ) 

Then when constructing the database, look for an entry in the accounts file of the form: “postgres_{host}_{database}”. Get the username and password, and plug them back into the DSN. 

Eventually, we’ll move the accounts file to another centralized key manager, or maybe, in production, into a configuration system like etcd for CoreOS deployments.  In anycase, we’ll have to have a way to store multiple credentials, for both test and production, in the same namespace. 

eric.
	


----------------------------------------------------------------------------------------------------------
Eric Busboom                                                                                (619) 363 2607        
eric@busboom.org                                                         http://www.busboom.org

**Comment by nmb10** [comment]
Database credentials from the accounts in my example are production credentials. We do not need test db credentials because I create new empty database by adding _test postfix to the database name. So, accounts do not need to keep two postgres sections for success testing. But, if we'll need two or more postres accounts, we can extend it with some postfix:

    postgres
        ...
    postgres-dev
        ...
    postgres-stage
        ...
and so on. Such values are easier to read and process. My suggestion for postgres credentials:

    postgres:
        host: 127.0.0.1
        database: ambry
        user: ambry
        password: secret

**Comment by ericbusboom** [comment]
> On Aug 1, 2015, at 10:31 PM, Kazbek <notifications@github.com> wrote:
> 
> Database credentials from the accounts in my example are production credentials. We do not need test db credentials because I create new empty database by adding _test postfix to the database name. So, accounts do not need to keep two postgres sections for success testing. But, if we'll need two or more postres accounts, we can extend it with some postfix:

Turns out most of the code for implementing multiple databases, and getting the credentials from the accounts file, was still in the code, left over from the earlier version. 

I updated the tests and test files to reflect this usage. See test/bundlefiles/ambry.yaml and test/bundlefiles/ambry-accounts.yaml for examples, and test.test_library.test_config.Test#test_database for usage. 

The library.database and library.warehouse configs can be either a DSN, or a name of a database in the database section of the config. In the database section, entries can be either a DSN, or a dict with broken out parts, as in your proposal. 

If a username and server is specified, but not a password, the RunConfig.database() method will use the accounts file to get the password. 

How about adding a database section to the config, with entries for ‘test-pg’ and ‘test-sqlite’. Then, when running tests, change the config for the library.database entry to either of those names. 

eric. 

----------------------------------------------------------------------------------------------------------
Eric Busboom                                                                                (619) 363 2607        
eric@busboom.org                                                         http://www.busboom.org
